### ISSUE TITLE
support for column vectors

### ISSUE THREAD

**Issue by Orbots** [mikera/core.matrix#217]
Right now vectors are row vectors.  In many domains column vectors are the norm.  Converting from clojure data structures to a column vectors (1xn matrix ) is awkward.  

The simplest way to make using column vectors easier would be if (transpose (matrix [1 2 3])) returned [[1][2][3]]

**Comment by mikera** [comment]
That behaviour would make `transpose` inconsistent: it has very well defined semantics at present that include keeping the same dimensionality.

On the other hand, you only need to add an extra pair of brackets to get what you want, so I suggest this as a solution:

``` clojure
(transpose (matrix [[1 2 3]]))
=> [[1] [2] [3]]
```

**Comment by Orbots** [comment]
That's not really a solution.  It's  still awkward.  The definition of the transpose of a row vector is that it becomes a column vector.  It seems concessions are being made to fit the design choices of the data structures rather than serving the actual thing that it is modeling.

**Comment by mars0i** [comment]
Orbots, actually, vectors are not unambiguously row vectors in core.matrix; it depends on context.  See the examples at the end of [this page](https://github.com/mikera/core.matrix/wiki/Vectors-vs.-matrices) in the wiki, and see the link at the bottom of the page, which is to a discussion in the Numerical Clojure Google group.  [This](https://groups.google.com/forum/#!searchin/numerical-clojure/$20Problems$20with$20transpose$3F$20/numerical-clojure/OOhkJ-azoYM/zJVutkoi1QkJ) discussion is also relevant.

**Comment by mikera** [comment]
@Orbots I think you may be getting confused over terminology.

In the array programming paradigm, a 1D vector is different from what you are calling a "row vector" or "column vector", both of which are actually 2D matrices (with one row and one column respectively).

a core.matrix vector doesn't have an orientation - it's just a vector. In the same way that a 1D array in most programming languages doesn't have a concept of orientation. `transpose` on a 1D vectors is just the identity function.

If you want "row vectors" and "column vectors" in the way that you are thinking, then you should represent them as 2D arrays (= matrices). In that case, `transpose` does what you expect.

Personally, I find these are less convenient to use, but a lot of people seem to like to think of everything as 2D matrices. Up to you - core.matrix supports both approaches

**Comment by mars0i** [comment]
Mikera's way of putting it is the correct.  The point I was making is that in some contexts, (1D) vectors are treated as if they are (2D) row or column vectors.

**Comment by Orbots** [comment]
Thanks for the all the info.  I've got it fairly clear now.  What I was implementing was a Conjugate Gradient algorithm, and then went on to some other similar algorithms.  There are lots of dot products between vectors.  Where dot products is r^t \* d, if r and d are Nx1 column vectors.  I was finding it pretty awkward to convert between clojure sequences to what I needed to do these dot products.  

I guess I was trying too hard!  I balked when I saw that (transpose v) -> v.  With my new knowledge I went back and tried (mmul (transpose (range 0 3)) (range 0 3))  this works as I would expect, isn't awkward and looks just like the source math would.  It's a bit subtle, and I'm not 100% comfortable with it but it's great that there is this nice concise way of doing what I want.  When in doubt probably best to stick with 2D matrices rather than vectors until I'm fully comfortable with the 1D vector paradigm.  

Thanks again!  I appreciate this library and it is working out really well for me.
